articles/ros_node_idl.md

```diff
@@ -0,0 +1,185 @@
+The interface may cover only a subset of nodes in a package, as long as the nodes that _are_ covered are done so completely.
+
+Here is an example IDL for a package containing two nodes:
+
+{% include_relative ros2_node_idl/interface_declaration.xml %}
+
+Once an IDL file is written, it is exported from the package manifest:
+
+{% include_relative ros2_node_idl/package.xml %}
+
+### Schema for `package.xml`'s export tag
+
+#### `interface`
+
+This is how the package exports its defined IDL for other tools to consume.
+
+Attributes:
+
+**path**:  Path to XML file containing IDL
+
+Note that the introduction of the `interface` tag within the `export` tag of the package manifest raises a small difficulty with regards to the [REP 149][rep149_export].
+The REP specifies:
+
+> To avoid potential collisions, an export tag should have the same name as the package which is meant to process it. The content of that tag is up to the package to define and use.
+
+Considering the high level abstraction of the IDL, the `interface` tag is not meant for a specific package but rather declares intrinsic properties for anyone to process it.
+However, the keyword is likely solidly descriptive enough to either be accepted as falling under [REP 149][rep149_export] or else to motivate an amendment of the REP.
+
+### IDL Schema
+
+#### `interface`
+
+This tag contains the package interface, made up of a collection of node interfaces.
+Root tag of the IDL file. There must be only one <interface> tag per IDL file.
+
+Attributes:
+- **version**: version of schema in use, allowing for future revisions.
+
+#### `node`
+
+Encapsulate a sequence of ROS entity interfaces.
+It is specific to a node as determined by its associated attributes.
+
+Attributes:
+- **name**: The base name of the node.
+
+#### `action`
+
+Define the interface for a given action.
+
+Attributes:
+- **name**: The name of the action.
+- **type**: The type of the action.
+Valid values are any ROS action types.
+- **server**: Whether or not the node provides an action server for the action.
+Valid values are "true" or "false". Defaults to "false".
+- **client**: Whether or not the node provides a client for the action.
+Valid values are "true" or "false". Defaults to "false".
+
+#### `parameter`
+
+Define the interface for a given parameter.
+
+Attributes:
+- **name**: The name of the parameter.
+- **type**: The type of the parameter.
+Valid values are any ROS parameter types.
+
+#### `service`
+
+Define the interface for a given service.
+
+Attributes:
+- **name**: The name of the service.
+- **type**: The type of the service.
+Valid values are any ROS service types.
+- **server**: Whether or not the node provides a server for the service.
+Valid values are "true" or "false". Defaults to "false".
+- **client**: Whether or not the node provides a client for the service.
+Valid values are "true" or "false". Defaults to "false".
+
+#### `topic`
+
+Define the interface for a given topic.
+
+Attributes:
```

Comment:
Since the QoS settings are deciding of two endpoints on the same topic name are being matched wouldn't it make sense to also specify those? Otherwise any "connectivity analysis" on these metadata might be incorrect.

Reply:
Fair point, going to update that.

Reply:
I added the QoS settings for actions, services and topics (da63aa62fabb216e04cd3bec6d35f54c33559e7f). I'm not familiar with the later two so please let me know your take on this.
I also went ahead assuming we would like those to defaults to the same values they do in code, hopefully I'm correctly linking to the defaults.

Reply:
Forgot to mention, I don't know what type should the `qos` attribute be. I'm not sure of to translate it to a xml tag attribute here...
If anyone has a suggestion?

Reply:
Given that they are objects in code, it may not be suited to being an attribute at all, but an object nested under the respective tags.

Reply:
You mean something similar to `<rosparam>` in `roslaunch`? 
E.g.
```xml
<topic name="/foo/bar" type="std_msgs/msg/String" subscription="true">
  <qos>
    policy_history: best_effort
    policy_reliabiliity: reliable
    ...
  </qos>
</topic>
```

Reply:
That's the right idea, yeah, although I'd make those QoS attributes, maybe something like:

```
<topic name="/foo/bar" type="std_msgs/msg/String" subscription="true">
  <qos policy_history="keep_last" policy_reliabiliity="reliable" \>
</topic>
```

Reply:
Updated `qos` so that it is a nested object in `action`, `service` and `topic`.
Updated the example to highlight `qos`. The example is not fully finished.